src/libraries/Common/src/SourceGenerators/TypeRef.cs

```diff
@@ -1,41 +1,47 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System;
 using System.Diagnostics;
 using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.DotnetRuntime.Extensions;
 
-namespace System.Text.Json.SourceGeneration
+namespace DotnetRuntime.SourceGenerators
 {
     /// <summary>
     /// An equatable value representing type identity.
     /// </summary>
-    [DebuggerDisplay("Name = {Name}")]
+    [DebuggerDisplay("Name={Name}")]
     public sealed class TypeRef : IEquatable<TypeRef>
     {
         public TypeRef(ITypeSymbol type)
         {
-            Name = type.Name;
-            FullyQualifiedName = type.GetFullyQualifiedName();
+            (Name, NamespaceDisplayString, FullyQualifiedDisplayString, MinimalDisplayString) = type.GetIdentifierInfo();
             IsValueType = type.IsValueType;
             TypeKind = type.TypeKind;
             SpecialType = type.SpecialType;
         }
 
         public string Name { get; }
 
-        /// <summary>
-        /// Fully qualified assembly name, prefixed with "global::", e.g. global::System.Numerics.BigInteger.
-        /// </summary>
-        public string FullyQualifiedName { get; }
+        public string NamespaceDisplayString { get; }
+
+        public string FullyQualifiedDisplayString { get; }
+
+        public string MinimalDisplayString { get; }
 
         public bool IsValueType { get; }
+
         public TypeKind TypeKind { get; }
+
         public SpecialType SpecialType { get; }
 
         public bool CanBeNull => !IsValueType || SpecialType is SpecialType.System_Nullable_T;
 
-        public bool Equals(TypeRef? other) => other != null && FullyQualifiedName == other.FullyQualifiedName;
+        public bool Equals(TypeRef? other) => other != null && FullyQualifiedDisplayString == other.FullyQualifiedDisplayString;
```

Comment:
This is not a valid equality check. Two `ITypeSymbols` can be different yet have the same fully qualified name.

Reply:
These types are being used in lieu of type symbols, which we have been advised to avoid caching in incremental values.

If source generators of this kind encounter two types with identical FQNs in the type graph they will most likely be generating ambiguous code. I believe it would result in the source generator failing at an earlier stage before any "ref" instances get created.

We should add unit testing validating that types with conflicting FQNs are being handled gracefully (i.e by emitting a diagnostic)

Reply:
IOW we needed a proxy for type equality that doesn't require encapsulating symbols.

Reply:
> If source generators of this kind encounter two types with identical FQNs in the type graph they will most likely be generating ambiguous code. 

Not necessarily. There are some cases where yes it would be ambiguous to emit them but there are other cases where it is not ambiguous. It's hard to enumerate all the cases but in short don't assume it's always ambiguous cause in many cases it's not. 

> IOW we needed a proxy for type equality that doesn't require encapsulating symbols.

For types in particular you can use AQN (assembly qualified names). Basically a combination of assembly name + FQN. Even that though is tricky because you need to consider generic argumens and apply it there as well.

Reply:
> For types in particular you can use AQN (assembly qualified names). Basically a combination of assembly name + FQN.

Guessing though that the same concerns apply when it comes to assembly equality? I.e. you could have two distinct assemblies with equal names, so an AQN representation can be ambiguous too?

Reply:
No. Compiler doesn't allow duplicate assembly names.

Reply:
i'm also skeptical of this approach.  Why is this information needed?  How is it used in the final generated source?

Reply:
The short answer is to avoid creating cycles in the source generator models when dealing with recursive types. Cycles aren't handled well from the perspective of record equality comparison, so a type "reference" needs to be used instead.

> Comparing types is problematic cause there are a lot of corner cases to consider (PIA, constraints, shape, change in structure, etc ...).

Agreed that this mode of comparison is not sound when considering type equality in the general case, however it does suffice when considering equality within the context of the particular SG models that type symbols project to: type graphs that don't contain any type parameters or NRT annotations (and can also be used as generic arguments themselves).

It's a deliberate compromise, I'd prefer it if we could use something OOTB that gets the right equality semantics for all C# types (including perhaps the possibility of encapsulating symbols in incremental values).